### PR TITLE
test(pacquet): complete Stage 10 long-tail ports

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Completes Stage 10 of pnpm/pnpm#13146 by porting the remaining long-tail pacquet suites for lockfile verification, optimistic repeat installs, catalogs, peer resolution, optional dependencies, install state, and hoisting.

Adds the missing optimistic repeat-install parity for ignored workspace-state settings and conflicted lockfiles. Features that are not implemented yet are represented by explicit known-failure boundaries in the port tracker.

This is Rust-only test/parity work. The corresponding TypeScript behavior is the upstream source coverage being ported.

## Squash Commit Body

```text
Port the remaining Stage 10 lockfile verification and optimistic repeat install coverage.

Add catalog, peer, optional dependency, install-state, and hoisting tests.

Record known-failure boundaries for features that are not implemented yet.

Related to pnpm/pnpm#13146.
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting.
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Keep it short and written for pnpm users — it becomes a release note.
- [x] Added or updated tests.
- [x] Updated the documentation if needed.

---
Written by an agent (Codex, GPT-5).

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/13161"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1787089262&installation_id=145934176&pr_number=13161&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F13161&signature=430247b5be62caea7ae4837f7fd82ad276a32336bd29e87603570e6184e01511"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevents optimistic repeat installs when lockfiles include merge-conflict markers.
  * Re-adding packages via dist-tags now preserves the original dependency group (e.g., stays in `devDependencies` when appropriate).
  * Improves catalog save/persistence and lockfile snapshot accuracy across add/install/update flows.
  * Makes hoisting and lifecycle bin creation more resilient when target paths already exist.
  * Strengthens lockfile verification and cache consistency, including minimumReleaseAge edge cases.
* **Documentation**
  * Updated the Rust test porting/coverage plan for newly handled scenarios.

### CODE REVIEW

File: pnpm/crates/package-manager/src/optimistic_repeat_install.rs [hunk 1]
```diff
@@ -316,6 +337,23 @@ pub fn check_optimistic_repeat_install(check: &OptimisticRepeatInstallCheck<'_>)
     }
 }
 
+fn has_conflicted_lockfile(check: &OptimisticRepeatInstallCheck<'_>) -> bool {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install.rs]: Addressed in c2a686c182. Changed lockfiles are now scanned in fixed-size chunks with a 16 MiB cap; unchanged regular lockfiles stay on the metadata-only path. Symlinks, non-regular files, oversized files, metadata/open/read failures, and files that grow past the cap conservatively fall back to the full install path. Added coverage for non-regular paths, a `/dev/zero` symlink, and oversized sparse lockfiles.

The same commit restores ordered short-circuit evaluation for workspace-state drift checks.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/package-manager/src/optimistic_repeat_install.rs [hunk 2]
```diff
@@ -138,6 +148,10 @@ pub fn check_optimistic_repeat_install(check: &OptimisticRepeatInstallCheck<'_>)
         return Decision::Skipped { reason: "optimistic_repeat_install disabled" };
     }
 
+    if has_conflicted_lockfile(check) {
+        return Decision::Skipped { reason: "a lockfile contains merge conflict markers" };
+    }
+
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/package-manager/src/optimistic_repeat_install.rs]: Addressed. Commit c2a686c182 moved conflict inspection after workspace-state loading and limits byte scanning to changed lockfiles. Commit dae0c57185 also applies the same check to `run`/`exec` dependency status, preserving the TypeScript message for actual conflict markers and conservatively reporting unsafe lockfiles.

Validation: 79 focused tests passed, plus Clippy, rustfmt, custom Dylint checks, and the full pre-push hook.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/tests/install_state.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    ))
+}
+
+fn external_lockfile_public_hoist() -> KnownResult<()> {
+    Err(KnownFailure::new(
+        "pacquet does not yet expose the external lockfile and project-root split required by the headless public-hoist scenario",
+    ))
+}
+
+#[test]
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/tests/install_state.rs:49]: No code change here. The working rules in pnpm/pnpm#13146 explicitly require either the upstream test or a `known_failures` stub at a not-yet-implemented boundary. These two tests document the existing PnP-without-symlinks and external-lockfile/public-hoist gaps; they do not wrap or suppress a failing implemented behavior. The passing install-state port in this file executes normally and asserts `.modules.yaml`, the current lockfile, and linked bins.

This uses the repository-wide `KnownFailure` mechanism already established for TEST_PORTING tracking, so removing the boundaries would lose the item-level record required by the issue.

---
Written by an agent (Codex, GPT-5).

**Comment by zkochan** [comment]
Addressed the final CodeRabbit style nit in 26ec0aad3a by logging each decision before the non-trivial assertion and simplifying the macro formatting. The focused optimistic-repeat-install suite passes (79 tests), and the full pre-push hook passed.

---
Written by an agent (Codex, GPT-5).
